hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/meta/SigTransactionMetadataBuilder.java

```diff
@@ -0,0 +1,238 @@
+    private ResponseCodeEnum status = OK;
+
+    private TransactionBody txn;
+    private final AccountKeyLookup keyLookup;
+    private AccountID payer;
+
+    public SigTransactionMetadataBuilder(@NonNull final AccountKeyLookup keyLookup) {
+        this.keyLookup = Objects.requireNonNull(keyLookup);
+    }
+
+    /**
+     * Sets status on {@link TransactionMetadata}. It will be {@link ResponseCodeEnum#OK} if there
+     * is no failure.
+     *
+     * @param status status to be set on {@link TransactionMetadata}
+     * @return builder object
+     */
+    public SigTransactionMetadataBuilder status(@NonNull final ResponseCodeEnum status) {
+        this.status = Objects.requireNonNull(status);
+        return this;
+    }
+
+    /**
+     * Set payer for the transaction. This is only used during {@link
+     * SigTransactionMetadata#copy(AccountKeyLookup)}
+     *
+     * @param payer payer for the transaction
+     * @return builder object
+     */
+    SigTransactionMetadataBuilder payer(@NonNull final AccountID payer) {
+        this.payer = Objects.requireNonNull(payer);
+        return this;
+    }
+
+    /**
+     * Add a keys to required keys list
+     *
+     * @param keys list of keys to add
+     * @return builder object
+     */
+    public SigTransactionMetadataBuilder addAllReqKeys(@NonNull final List<HederaKey> keys) {
+        requiredKeys.addAll(Objects.requireNonNull(keys));
+        return this;
+    }
+
+    /**
+     * Fetches the payer key and add to required keys in {@link TransactionMetadata}.
+     *
+     * @param payer payer for the transaction
+     * @return builder object
+     */
+    public SigTransactionMetadataBuilder payerKeyFor(@NonNull AccountID payer) {
+        this.payer = Objects.requireNonNull(payer);
+        addPayerKey();
+        return this;
+    }
+
+    /**
+     * Adds given key to required keys in {@link TransactionMetadata}.
+     *
+     * @param key key to be added
+     * @return builder object
+     */
+    public SigTransactionMetadataBuilder addToReqKeys(@NonNull HederaKey key) {
+        requiredKeys.add(Objects.requireNonNull(key));
+        return this;
+    }
+
+    /**
+     * Adds the {@link TransactionBody} of the transaction on {@link TransactionMetadata}.
+     *
+     * @param txn transaction body of the transaction
+     * @return builder object
+     */
+    public SigTransactionMetadataBuilder txnBody(@NonNull TransactionBody txn) {
+        this.txn = Objects.requireNonNull(txn);
+        return this;
+    }
+
+    /**
+     * Checks if the accountId is same as payer or the status of the metadata is already failed. If
+     * either of the above is true, doesn't look up the keys for given account. Else, looks up the
+     * keys for account. If the lookup fails, sets the default failureReason given in the result.
+     *
+     * @param id given accountId
+     */
+    public SigTransactionMetadataBuilder addNonPayerKey(@NonNull final AccountID id) {
```

Comment:
I don't think there should be any `add*Key()` method that bypasses the `isNotNeeded() -> addToKeysOrFail()` pattern.